docs/design/2020-07-07-change-column-types.md

```diff
@@ -0,0 +1,105 @@
+2. In the reorganization processing stage, row data and index data need to be processed in batches, which can be similar to the `add index` operation during initial implementation. Get the batch processing range, and then use the `newType` to construct the `changingColA` and `changingIdxA` information for the insert operation (this operation requires the corresponding row and index column to be locked). If errors such as data truncated are encountered at this stage, it is necessary to report the errors and roll back to exit.
+3. The next phase of change requires 3 status changes:
+    * Lock this table and cannot write to it.
+    * Drop `colA` and `idxA`, change the status of `changingColA` and `changingIdxA` to public, and change their names, offsets, etc.
+    * Unlock this table.
+
+### Note
+
+This operation itself is an `alter table tableName modify/change column …` statement, which includes changing the column type, changing the column attribute from null to not null, changing the column offset, etc. Considering the atomicity, some operations need to be changed together.
+
+In addition, due to the complexity of this feature, considering that the friendly review needs to be divided into multiple PRs, a switch can be added to deal with the problem of multiple PRs.
+
+### Rolling back
+
+This requires changing the state of `colA` and `idxA` to `StatePublic`, and then dropping `changingColA` and `changingIdxA`. Of course, if you modify the flag and other attributes, you need to roll back.
+
+## Compatibility
+
+### Compatibility issues with MySQL
+
+* Considering the complexity of supporting clustered-index, and the current TiDB does not fully support clustered-index, so temporarily does not support column with primary key for type modification.
+
+### Compatibility issues with TiDB
+
+The DDL statement itself already has a corresponding type (ActionModifyColumn). If you need to improve this operation, you will encounter compatibility issues when rolling upgrades. This problem is handled by adding a global variable `tidb_enable_change_column_type`. When this value is true, the `modify/change column` statement is allowed to perform more types of column type changes (features supported by this proposal).
+
+* For new clusters, set this variable to true during bootstrap.
+* For clusters requiring rolling upgrade, the default value is false, and users need to explicitly set this variable to true.
+
+### Compatible with other components
+
+It needs to be synchronized this feature with components such as Tools, TiFlash and BR.
+
+## Implementation
+
+In addition to considering the update of the corresponding column type and data, and the update of the data in the index column involved, the modification of the column type operation also needs to consider some related characteristics as follows:
+
+### Generate columns
+
+When the modified type column has a related generated column (whether stored or virtual), it has the following characteristics:
```

Comment:
```suggestion
When the modified type relates to generated columns (either stored or virtual), the implementation has the following characteristics:
```

Reply:
We have better make it as TODO since we don't support it in the primary implementation of CTC.